src/main/kotlin/com/procurify/flagcounter/FlagCounter.kt

```diff
@@ -1,24 +1,154 @@
+                            }
+                        }
+                    }
+                    if (failedTeamMessages.size > 0) {
+                        errorMessager.postOrThrow(
+                                "Failed to send updates for ${
+                                    failedTeamMessages.joinToString(separator = "\n") { it.email }
+                                }}"
+                        )
+                    }
+                }
+        )
+    }
+
+    /**
+     * Groups the [FlagDetail] in a list as values to a map keyed by the provided [teamEmails]. If there are maintainers
+     * returned in the [FlagResponse] that are not in the provided [teamEmails], those flags are not added to the map
+     * TODO There is likely a cleaner way to do this without a mutable map
+     */
+    fun groupFlagsByMaintainer(
+            flagDetails: List<FlagDetailAndStatus>,
+            teamEmails: Set<TeamEmail>
+    ): Map<FlagMaintainer, List<FlagDetailAndStatus>> = flagDetails
+            .fold(mapOf<FlagMaintainer, List<FlagDetailAndStatus>>().toMutableMap()) { acc, flagDetail ->
+                acc[flagDetail._maintainer] = acc[flagDetail._maintainer].orEmpty() + flagDetail
+                acc
+            }
+            .filter { teamEmails.contains(TeamEmail(it.key.email)) }
+
+    /**
+     * Post a message to the [Messager] and throw an exception with the same [message] if the post fails
+     */
+    private fun Messager.postOrThrow(message: String) =
+            this.postMessage(message).mapLeft { throw Exception(message) }
+
+    /**
+     * Formats a [totalFlagCount] into a message to be sent by the messager
+     */
+    private fun formatTotalCountMessage(totalFlagCount: Int): String {
+        val comparisonMessage = FlagEquivalentMessageGenerator.getNearestNumberMessage(totalFlagCount)
+        return "There are currently $totalFlagCount flags in the system!\n$comparisonMessage"
+    }
+
+    private fun formatTeamMessage(team: FlagMaintainer, teamFlags: List<FlagDetailAndStatus>): String {
+        val removeCount = teamFlags.filter { it.status in setOf(Status.INACTIVE, Status.LAUNCHED) }.size
+        // TODO Parameterize link url based on project/environment configuration
+        return """Hey ${team.firstName}!
+           |Launch Darkly thinks $removeCount of your ${teamFlags.size} flags could be ready for removal.
+           |Take a look https://app.launchdarkly.com/default/production/features""".trimMargin()
+    }
+
+    data class FlagDetailAndStatus(
+            val key: String,
+            val _maintainer: FlagMaintainer,
+            val status: Status
+    )
 }
 
+/**
+ * Identifier for a Team
+ */
+data class TeamEmail(
+        val email: String
+)
+
 interface Messager {
-    fun postMessage(message: String)
+    fun postMessage(message: String): Either<MessagerError, Unit>
+
+    data class MessagerError(
+            val message: String
+    )
 }
 
 interface FlagReader {
```

Comment:
Not sure if this is a Kotlin standard, but I would personally lean towards putting interfaces and types at the top of the file. I assume this wouldn't make much of a difference during regular development workflows, but I find it's beneficial for someone coming into a file or reviewing code like this to be able to get a frame of reference for these things prior to reading further into the code that makes use of or implements them.

Reply:
Interesting, I think I must be doing it as a force of habit because I don't see any prescription in the Kotlin style guide. I think I will just move them into different files